source/custom/rules/pci_id_rule.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rules
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+
+	"sigs.k8s.io/node-feature-discovery/source/internal"
+)
+
+// Rule that matches on PCI the following PCI device attributes: <class, vendor, device>
+// each device attribute will be composed of list elements.
+// Match operation: OR will be performed per element and AND will be performed per attribute.
+// An empty attribute will not be included in the matching process.
+type pciIdRuleInput struct {
+	Class  []string `json:"class,omitempty"`
+	Vendor []string `json:"vendor,omitempty"`
+	Device []string `json:"device,omitempty"`
+}
+
+type PciIdRule struct {
+}
+
+// Match PCI devices on the provided list of PCI regular expressions
+func (r PciIdRule) Match(input []byte) (bool, error) {
+	pciId := pciIdRuleInput{}
+	err := json.Unmarshal(input, &pciId)
+	if err != nil {
+		return false, fmt.Errorf("failed to convert raw JSON input: %s", err.Error())
+	}
+
+	log.Printf("INFO: PCI ID Rule: Match() with: %+v", pciId)
+	devAttr := map[string]bool{}
+	for _, attr := range []string{"class", "vendor", "device"} {
+		devAttr[attr] = true
+	}
+	allDevs, err := pciutils.DetectPci(devAttr)
+	if err != nil {
+		return false, fmt.Errorf("failed to detect PCI devices: %s", err.Error())
+	}
+
+	for _, classDevs := range allDevs {
+		for _, dev := range classDevs {
+			// match rule on a single device
+			if r.matchDevOnRule(dev, pciId) {
+				log.Printf("INFO: Rule Match on device: %v", dev)
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+func (r PciIdRule) matchDevOnRule(dev pciutils.PciDeviceInfo, rule pciIdRuleInput) bool {
+	if len(rule.Class) == 0 && len(rule.Vendor) == 0 && len(rule.Device) == 0 {
+		return false
+	}
+
+	if len(rule.Class) == 0 || in(dev["class"], rule.Class) == true {
+		if len(rule.Vendor) == 0 || in(dev["vendor"], rule.Vendor) == true {
+			if len(rule.Device) == 0 || in(dev["device"], rule.Device) == true {
```

Comment:
nit: you can drop the `== true` from the conditionals

suggestion: to make this more readable (especially in the case of future additions) we could drop the nested `if`s by doing smth like:
```
if len(rule.Class) > 0 && !in(dev["class"], rule.Class) {
    return false
}
```